cmd/servicelog/post.go

```diff
@@ -24,175 +26,208 @@ import (
 	"github.com/spf13/cobra"
 )
 
-var (
+type PostCmdOptions struct {
 	Message         servicelog.Message
 	ClustersFile    servicelog.ClustersFile
-	template        string
+	Template        string
 	filterFiles     []string // Path to filter file
 	filtersFromFile string   // Contents of filterFiles
 	isDryRun        bool
 	skipPrompts     bool
 	clustersFile    string
 	internalOnly    bool
+	Args            []string
```

Comment:
I'm -1 to this. If we're going the `Init()` approach it makes more sense to me for it to be something explicit like this:

```go
func (p *PostCmdOptions) Init(clusterId, template string) error {
  p.template = template
  p.clusterId = clusterId
  // ...etc

  return nil
}

Reply:
The issue with that approach is that we can't hold things to an interface if we accept random parameters for each command.

 Really, the only point of the `Init()` method is to be able to instantiate default sdks that we might want to mock.  
 
 So if a command needs both the ocm and aws clients, the Init would look something like this:
 
 ```
 func (p *PostCmdOptions) Init() error {
   if p.ocmClient == nil {
     p.ocmClient = ocm.newOcmClient()
   }
   
   if p.awsClient == nil {
     p.awsClient = aws.newAwsClient()
   }
}
```

Ideally so that when you want to run this command from a unit test you would do something like
```
postCmd := newPostCmdOptions()
postCmd.awsClient = mockAwsClient.EXPECT()...
postCmd.ocmClient = mockOcmClient.EXPECT()...

postCmd.Run()
```

So the point of `Init` is to be able to ensure that when we run the commands via the CLI invocation that we have the actual OCM/AWS clients vs when we're testing we can push this in.

I'm not terribly convinced this is the greatest name, so if you have any other ideas around the convention to use here then I'm all ears.

Reply:
A part of me wants to rewrite this entire command to not use an interface at all, but be more in line with what we did for `checkbanneduser.go` and `validatepullsecret.go`. For external context, the functionality executed by those commands is in public functions that are not tied to any struct. Their dependencies are defined as function parameters, and they can be called from anywhere.

Do you think there's value in writing up an alternative PR proposal with that approach to compare how it looks?

Reply:
Ah I brought it up mostly because it seems like a bit of a gap in this specific use-case, I can see Kirk's point where `Init()` may be useful in other situations. My suggestion doesn't need to overwrite the intended use-case for `Init()`, but the core part of my gripe is I think we should have a clearer way of setting up a `PostCmdOptions` struct than making the caller pass a cluster id via the args field.

Reply:
The "args" can only ever be of length 0 or 1. How about instead of having the "Args" field at all we just make it `clusterId string` and it can either be empty or have a value? I think I'm going to go with that approach unless there are objections.

Reply:
Agreed, because we're essentially not caring about the additional functionality of the servicelog post command with regards to `filterParams` and `clustersFile`, so we can make a clear interface for sending servicelogs to one cluster id.

(i.e. this check in the code)
```go
if o.ClusterId == "" && len(filterParams) == 0 && o.clustersFile == ""
```

Reply:
I'm going to throw "Validate" and "init" calls at the start of "Run" for now to make sure they _can't_ get skipped over. They're idempotent functions and very lightweight so calling them multiple times shouldn't cause any harm.